p2p/protocol/identify/id.go

```diff
@@ -178,112 +201,115 @@ func NewIDService(h host.Host, opts ...Option) (*idService, error) {
-				ph, ok := phs[rp]
-				if !ok {
-					// move on, move on, there's nothing to see here.
-					continue
-				}
-				// This is idempotent if already stopped.
-				ph.stop()
-			}
-
-		case rp := <-phClosedCh:
-			ph := phs[rp]
-
-			// If we are connected to the peer, it means that we got a connection from the peer
-			// before we could finish removing it's handler on the previous disconnection.
-			// If we delete the handler, we wont be able to push updates to it
-			// till we see a new connection. So, we should restart the handler.
-			// The fact that we got the handler on this channel means that it's context and handler
-			// have completed because we write the handler to this chanel only after it closed.
-			if ids.Host.Network().Connectedness(rp) == network.Connected {
-				ph.start(handlerCtx, func() { phClosedCh <- rp })
-			} else {
-				delete(phs, rp)
-			}
-
+		case <-ids.triggerPush:
+			ids.refCount.Add(1)
+			go func() {
+				defer ids.refCount.Done()
+				ids.sendPushes(ctx)
+			}()
 		case e, more := <-sub.Out():
 			if !more {
 				return
 			}
+			ids.updateSnapshot()
 			switch e.(type) {
-			case event.EvtLocalAddressesUpdated:
-				for pid := range phs {
-					select {
-					case phs[pid].pushCh <- struct{}{}:
-					default:
-						log.Debugf("dropping addr updated message for %s as buffer full", pid)
-					}
-				}
-			case event.EvtLocalProtocolsUpdated:
-				for pid := range phs {
-					select {
-					case phs[pid].pushCh <- struct{}{}:
-					default:
-						log.Debugf("dropping protocol updated message for %s as buffer full", pid)
-					}
+			case event.EvtLocalAddressesUpdated, event.EvtLocalProtocolsUpdated:
+				// trigger a push
+				select {
+				case ids.triggerPush <- struct{}{}:
+				default: // another push is already queued
 				}
 			}
-		case <-ids.ctx.Done():
+		case <-ctx.Done():
 			return
 		}
 	}
 }
 
+func (ids *idService) sendPushes(ctx context.Context) {
+	select {
+	case ids.pushSemaphore <- struct{}{}:
+	default:
+		// another sendPushes call is currently running
+		return
+	}
+	defer func() { <-ids.pushSemaphore }()
+
+	ids.connsMu.RLock()
+	conns := make([]network.Conn, 0, len(ids.conns))
+	for c, e := range ids.conns {
+		// Push even if we don't know if push is supported.
+		// This will be only the case while the IdentifyWaitChan call is in flight.
+		if e.PushSupport == identifyPushSupported || e.PushSupport == identifyPushSupportUnknown {
+			conns = append(conns, c)
+		}
+	}
+	ids.connsMu.RUnlock()
+
+	sem := make(chan struct{}, maxPushConcurrency)
+	for _, c := range conns {
+		// check if the connection is still alive
+		ids.connsMu.RLock()
+		e, ok := ids.conns[c]
+		ids.connsMu.RUnlock()
+		if !ok {
+			continue
+		}
+		// check if we already sent the current snapshot to this peer
+		ids.currentSnapshot.Lock()
+		snapshot := ids.currentSnapshot.snapshot
+		ids.currentSnapshot.Unlock()
+		if !e.Timestamp.Before(snapshot.timestamp) {
```

Comment:
Who updates `e.Timestamp`?

Reply:
This is now done in `sendIdentifyResp`. Not sure how to properly test it though.

Reply:
A unit test that makes sure we have the timestamp in the connection after doing an identify?

Reply:
I'd prefer to test behavior, not implementation specifics.